scipy/special/_logsumexp.py

```diff
@@ -100,53 +101,113 @@ def logsumexp(a, axis=None, b=None, keepdims=False, return_sign=False):
+        sgn = xp.sign(out)
+    else:
+        with np.errstate(divide='ignore', invalid='ignore'):  # log of zero is OK
+            out, sgn = _logsumexp(a, b, axis=axis, return_sign=return_sign, xp=xp)
+
+    # Deal with shape details - reducing dimensions and convert 0-D to scalar for NumPy
+    out = xp.squeeze(out, axis=axis) if not keepdims else out
+    sgn = xp.squeeze(sgn, axis=axis) if (sgn is not None and not keepdims) else sgn
+    out = out[()] if out.ndim == 0 else out
+    sgn = sgn[()] if (sgn is not None and sgn.ndim == 0) else sgn
+
+    return (out, sgn) if return_sign else out
+
+
+def _elements_and_indices_with_max_real(a, axis=-1, xp=None):
+    # We need *an* element with the maximum real part.
+    # If the array is complex, we need to choose *one* of the complex values,
+    # not all values with the same real part.
+    # We could use `argmax` on real component, but array API doesn't yet have
+    # `take_along_axis`, and even if it did, we would have problems with axis tuples.
+    xp = array_namespace(a) if xp is None else xp
+
+    if xp.isdtype(a.dtype, "complex floating"):
+        # select all elements with max real part.
+        max = xp.max(xp.real(a), axis=axis, keepdims=True)
+        mask = xp.real(a) == max
+
+        # Of those, choose one arbitrarily. This is a reasonably
+        # simple, array-API compatible way of doing so that doesn't
+        # have a problem with `axis` being a tuple or None.
+        i = xp.reshape(xp.arange(xp_size(a)), a.shape)
+        i[~mask] = -1
+        max_i = xp.max(i, axis=axis, keepdims=True)
+        mask = i == max_i
+        max = xp.reshape(a[mask], max_i.shape)
     else:
-        real = xp.real(a) if xp.isdtype(a.dtype, "complex floating") else a
-        a_max = xp.max(real, axis=axis, keepdims=True)
+        max = xp.max(a, axis=axis, keepdims=True)
+        mask = a == max
 
-    if a_max.ndim > 0:
-        a_max[~xp.isfinite(a_max)] = 0
-    elif not xp.isfinite(a_max):
-        a_max = 0
+    return xp.asarray(max), xp.asarray(mask)
 
-    if b is not None:
-        b = xp.asarray(b)
-        tmp = b * xp.exp(a - a_max)
-    else:
-        tmp = xp.exp(a - a_max)
 
-    # suppress warnings about log of zero
-    with np.errstate(divide='ignore'):
-        s = xp.sum(tmp, axis=axis, keepdims=keepdims, dtype=tmp.dtype)
-        if return_sign:
-            # For complex, use the numpy>=2.0 convention for sign.
-            if xp.isdtype(s.dtype, "complex floating"):
-                sgn = s / xp.where(s == 0, xp.asarray(1, dtype=s.dtype), xp.abs(s))
-            else:
-                sgn = xp.sign(s)
-            s = xp.abs(s)
-        out = xp.log(s)
-
-    if not keepdims:
-        a_max = xp.squeeze(a_max, axis=axis)
-    out += a_max
+def _sign(x, xp):
+    return x / xp.where(x == 0, xp.asarray(1, dtype=x.dtype), xp.abs(x))
+
 
+def _logsumexp(a, b, axis, return_sign, xp):
+
+    # This has been around for about a decade, so let's consider it a feature:
```

Comment:
Seems OK, it's what I'd expect intuitively :)

Reply:
0 * np.inf = 0?

Reply:
That is undefined, of course, but I think the general expectation when assigning weights zero is that those values disappear. But, as you say, it's already established, so nothing to do here.